File: app/src/main/java/net/bible/service/db/mynote/MyNoteDBAdapter.java
/**
 * 
 */
package net.bible.service.db.mynote;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.SQLException;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteException;
import android.database.sqlite.SQLiteOpenHelper;
import android.util.Log;

import net.bible.service.db.CommonDatabaseHelper;
import net.bible.service.db.bookmark.BookmarkDatabaseDefinition.BookmarkColumn;
import net.bible.service.db.mynote.MyNoteDatabaseDefinition.MyNoteColumn;
import net.bible.service.db.mynote.MyNoteDatabaseDefinition.Table;

import org.apache.commons.lang3.StringUtils;
import org.crosswire.jsword.passage.Key;
import org.crosswire.jsword.passage.NoSuchKeyException;
import org.crosswire.jsword.passage.VerseKey;
import org.crosswire.jsword.passage.VerseRange;
import org.crosswire.jsword.passage.VerseRangeFactory;
import org.crosswire.jsword.versification.BibleBook;
import org.crosswire.jsword.versification.Versification;
import org.crosswire.jsword.versification.system.Versifications;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

/**
 * MyNote database update methods
 * 
 * @see gnu.lgpl.License for license details.<br>
 *      The copyright to this program is held by it's authors.
 * @author John D. Lewis [balinjdl at gmail dot com]
 * @author Martin Denham [mjdenham at gmail dot com]
 */
public class MyNoteDBAdapter {

	// Variable to hold the database instance
	private SQLiteDatabase db;

	// Database open/upgrade helper
	private SQLiteOpenHelper dbHelper;
	
	private static final String TAG = "MyNoteDBAdapter";

	public MyNoteDBAdapter() {
		dbHelper =  CommonDatabaseHelper.getInstance(); 
	}

	public MyNoteDBAdapter open() throws SQLException {
		try {
			Log.d(TAG, "about to getWritableDatabase");
			db = dbHelper.getWritableDatabase();
		} catch (SQLiteException ex) {
			Log.d(TAG, "about to getReadableDatabase");
			db = dbHelper.getReadableDatabase();
		}
		return this;
	}

	public void close() {
		//TODO remove totally or call on app stop
		// apparently we are now supposed to leave the database open and allow Android to close it when appropriate
		// db.close();
	}

	public MyNoteDto insertMyNote(MyNoteDto mynote) {
		// Create a new row of values to insert.
		Log.d(TAG, "about to insertMyNote: " + mynote.getVerseRange());
		VerseRange verseRange = mynote.getVerseRange();
		String v11nName = getVersification(verseRange);
        // Gets the current system time in milliseconds
        Long now = Long.valueOf(System.currentTimeMillis());

		ContentValues newValues = new ContentValues();
		newValues.put(MyNoteColumn.KEY, verseRange.getOsisRef());
		newValues.put(MyNoteColumn.VERSIFICATION, v11nName);
		newValues.put(MyNoteColumn.MYNOTE, mynote.getNoteText());
		newValues.put(MyNoteColumn.LAST_UPDATED_ON, now);
		newValues.put(MyNoteColumn.CREATED_ON, now);
		
		long newId = db.insert(Table.MYNOTE, null, newValues);
		MyNoteDto newMyNote = getMyNoteDto(newId);
		return newMyNote;
	}

	/**
	 * @param key
	 * @return
	 */
	private String getVersification(Key key) {
		String v11nName="";
		if (key instanceof VerseKey) {
			// must save a VerseKey's versification along with the key!
			v11nName = ((VerseKey<?>) key).getVersification().getName();
		}
		return v11nName;
	}

	public MyNoteDto updateMyNote(MyNoteDto mynote) {
		// Create a new row of values to insert.
		Log.d(TAG, "about to updateMyNote: " + mynote.getVerseRange());
		VerseRange verserange = mynote.getVerseRange();
		String v11nName = getVersification(verserange);
        // Gets the current system time in milliseconds
        Long now = Long.valueOf(System.currentTimeMillis());

		ContentValues newValues = new ContentValues();
		newValues.put(MyNoteColumn.KEY, verserange.getOsisRef());
		newValues.put(MyNoteColumn.VERSIFICATION, v11nName);
		newValues.put(MyNoteColumn.MYNOTE, mynote.getNoteText());
		newValues.put(MyNoteColumn.LAST_UPDATED_ON, now);
		
		long rowsUpdated = db.update(Table.MYNOTE, newValues, "_id=?", new String []{String.valueOf(mynote.getId())});
		Log.d(TAG, "Rows updated:"+rowsUpdated);
		
		return getMyNoteDto(mynote.getId());
	}

	public boolean removeMyNote(MyNoteDto mynote) {
		Log.d(TAG, "Removing my note:" + mynote.getVerseRange());
		return db.delete(Table.MYNOTE, MyNoteColumn._ID + "=" + mynote.getId(), null) > 0;
	}

	public List<MyNoteDto> getAllMyNotes() {
		Log.d(TAG, "about to getAllMyNotes");
		List<MyNoteDto> allMyNotes = new ArrayList<MyNoteDto>();
		Cursor c = db.query(MyNoteQuery.TABLE, MyNoteQuery.COLUMNS, null, null, null, null, null);
		try {
			if (c.moveToFirst()) {
		        while (!c.isAfterLast()) {
		        	MyNoteDto mynote = getMyNoteDto(c);
		    		allMyNotes.add(mynote);
		       	    c.moveToNext();
		        }
			}
		} finally {
	        c.close();
		}
        
		Log.d(TAG, "allMyNotes set to " + allMyNotes.size() + " item long list");
        return allMyNotes;
	}
	
	public List<MyNoteDto> getMyNotesInBook(BibleBook book) {
		Log.d(TAG, "about to getMyNotesInPassage:"+book.getOSIS());
		List<MyNoteDto> notesList = new ArrayList<MyNoteDto>();
		Cursor c = db.query(MyNoteQuery.TABLE, MyNoteQuery.COLUMNS, MyNoteColumn.KEY+" LIKE ?", new String []{String.valueOf(book.getOSIS()+".%")}, null, null, null);
		try {
			if (c.moveToFirst()) {
		        while (!c.isAfterLast()) {
		        	MyNoteDto mynote = getMyNoteDto(c);
		    		notesList.add(mynote);
		       	    c.moveToNext();
		        }
			}
		} finally {
	        c.close();
		}
        
		Log.d(TAG, "myNotesInPassage set to " + notesList.size() + " item long list");
        return notesList;
	}

	public MyNoteDto getMyNoteDto(long id) {
		MyNoteDto mynote = null;
		
		Cursor c = db.query(MyNoteQuery.TABLE, MyNoteQuery.COLUMNS, MyNoteColumn._ID+"=?", new String[] {String.valueOf(id)}, null, null, null);
		try {
			if (c.moveToFirst()) {
				mynote = getMyNoteDto(c);
			}
		} finally {
			c.close();
		}
		
		return mynote;
	}

	/**
	 * Find bookmark starting at the location specified by key.
	 * If it starts there then the initial part of the key should match.
	 */
	public MyNoteDto getMyNoteByStartVerse(String startVerse) {
		MyNoteDto mynote = null;
		Cursor c=null;
		try {
			// exact match
			c = db.query(MyNoteQuery.TABLE, MyNoteQuery.COLUMNS, MyNoteColumn.KEY+"=?", new String[] {startVerse}, null, null, null);
			if (!c.moveToFirst()) {
				// start of verse range
				c = db.query(MyNoteQuery.TABLE, MyNoteQuery.COLUMNS, MyNoteColumn.KEY + " LIKE ?", new String[]{startVerse + "-%"}, null, null, null);
				if (!c.moveToFirst()) {
					return null;
				}
			}
			mynote = getMyNoteDto(c);
		} finally {
			c.close();
		}

		return mynote;
	}

	/** return Dto from current cursor position or null
	 * @param c
	 * @return
	 * @throws NoSuchKeyException
	 */
	private MyNoteDto getMyNoteDto(Cursor c) {
		MyNoteDto dto = new MyNoteDto();
		try {
			//Id
			Long id = c.getLong(MyNoteQuery.ID);
			dto.setId(id);
			
			//Verse
			String key = c.getString(MyNoteQuery.KEY);
			Versification v11n=null;
			if (!c.isNull(MyNoteQuery.VERSIFICATION)) {
				String v11nString = c.getString(MyNoteQuery.VERSIFICATION);
				if (!StringUtils.isEmpty(v11nString)) {
					v11n = Versifications.instance().getVersification(v11nString);
				}
			}
			if (v11n==null) {
				Log.d(TAG, "Using default Versification");
				// use default v11n
				v11n = Versifications.instance().getVersification(Versifications.DEFAULT_V11N);
			}
			Log.d(TAG, "Versification found:"+v11n);
			try {
				dto.setVerseRange(VerseRangeFactory.fromString(v11n, key));
			} catch (Exception e) {
				Log.e(TAG, "Note saved with incorrect versification", e);
				// fix problem where KJV was always the v11n even for dc books
				// NRSVA should contain most dc books and allow verse to be fetched
				Versification v11nWithDC = Versifications.instance().getVersification("NRSVA");
				dto.setVerseRange(VerseRangeFactory.fromString(v11nWithDC, key));
			}

			//Note
			String mynote = c.getString(MyNoteQuery.MYNOTE);
			dto.setNoteText(mynote);
			
			//Update date
			long updated = c.getLong(MyNoteQuery.LAST_UPDATED_ON);
			dto.setLastUpdatedOn(new Date(updated));

			//Create date
			long created = c.getLong(MyNoteQuery.CREATED_ON);
			dto.setCreatedOn(new Date(created));
			
		} catch (NoSuchKeyException nke) {
			Log.e(TAG, "Key error", nke);
		}
		
		return dto;
	}
	
	private interface MyNoteQuery {
        String TABLE = Table.MYNOTE;

		String[] COLUMNS = new String[] {MyNoteColumn._ID, MyNoteColumn.KEY, BookmarkColumn.VERSIFICATION, MyNoteColumn.MYNOTE, MyNoteColumn.LAST_UPDATED_ON, MyNoteColumn.CREATED_ON};

        int ID = 0;
        int KEY = 1;
        int VERSIFICATION = 2;
        int MYNOTE = 3;
        int LAST_UPDATED_ON = 4;
        int CREATED_ON = 5;
    }	
}

File: app/src/main/java/net/bible/service/db/mynote/MyNoteDBAdapter.kt
/*
 * Copyright (c) 2020 Martin Denham, Tuomas Airaksinen and the And Bible contributors.
 *
 * This file is part of And Bible (http://github.com/AndBible/and-bible).
 *
 * And Bible is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software Foundation,
 * either version 3 of the License, or (at your option) any later version.
 *
 * And Bible is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with And Bible.
 * If not, see http://www.gnu.org/licenses/.
 *
 */
/**
 *
 */
package net.bible.service.db.mynote

import android.content.ContentValues
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase.CONFLICT_FAIL
import android.util.Log
import androidx.sqlite.db.SupportSQLiteDatabase
import androidx.sqlite.db.SupportSQLiteQueryBuilder.builder
import net.bible.service.db.DatabaseContainer
import net.bible.service.db.bookmark.BookmarkDatabaseDefinition.BookmarkColumn
import net.bible.service.db.mynote.MyNoteDatabaseDefinition.MyNoteColumn
import org.apache.commons.lang3.StringUtils
import org.crosswire.jsword.passage.Key
import org.crosswire.jsword.passage.NoSuchKeyException
import org.crosswire.jsword.passage.VerseKey
import org.crosswire.jsword.passage.VerseRangeFactory
import org.crosswire.jsword.versification.BibleBook
import org.crosswire.jsword.versification.Versification
import org.crosswire.jsword.versification.system.Versifications
import java.util.*

/**
 * MyNote database update methods
 *
 * @author John D. Lewis [balinjdl at gmail dot com]
 * @author Martin Denham [mjdenham at gmail dot com]
 */
class MyNoteDBAdapter {
    // Variable to hold the database instance
	private val db: SupportSQLiteDatabase = DatabaseContainer.db.openHelper.readableDatabase

    fun insertMyNote(mynote: MyNoteDto): MyNoteDto? { // Create a new row of values to insert.
        Log.d(TAG, "about to insertMyNote: " + mynote.verseRange)
        val verseRange = mynote.verseRange
        val v11nName = getVersification(verseRange)
        // Gets the current system time in milliseconds
        val now = java.lang.Long.valueOf(System.currentTimeMillis())
        val newValues = ContentValues()
        newValues.put(MyNoteColumn.KEY, verseRange.osisRef)
        newValues.put(MyNoteColumn.VERSIFICATION, v11nName)
        newValues.put(MyNoteColumn.MYNOTE, mynote.noteText)
        newValues.put(MyNoteColumn.LAST_UPDATED_ON, now)
        newValues.put(MyNoteColumn.CREATED_ON, now)
        val newId = db.insert(MyNoteDatabaseDefinition.Table.MYNOTE, CONFLICT_FAIL, newValues)
        return getMyNoteDto(newId)
    }

    /**
     * @param key
     * @return
     */
    private fun getVersification(key: Key): String {
        var v11nName = ""
        if (key is VerseKey<*>) { // must save a VerseKey's versification along with the key!
            v11nName = key.versification.name
        }
        return v11nName
    }

    fun updateMyNote(mynote: MyNoteDto): MyNoteDto? { // Create a new row of values to insert.
        Log.d(TAG, "about to updateMyNote: " + mynote.verseRange)
        val verserange = mynote.verseRange
        val v11nName = getVersification(verserange)
        // Gets the current system time in milliseconds
        val now = java.lang.Long.valueOf(System.currentTimeMillis())
        val newValues = ContentValues()
        newValues.put(MyNoteColumn.KEY, verserange.osisRef)
        newValues.put(MyNoteColumn.VERSIFICATION, v11nName)
        newValues.put(MyNoteColumn.MYNOTE, mynote.noteText)
        newValues.put(MyNoteColumn.LAST_UPDATED_ON, now)
        val rowsUpdated = db.update(MyNoteDatabaseDefinition.Table.MYNOTE, CONFLICT_FAIL, newValues, "_id=?", arrayOf(mynote.id.toString())).toLong()
        Log.d(TAG, "Rows updated:$rowsUpdated")
        return getMyNoteDto(mynote.id!!)
    }

    fun removeMyNote(mynote: MyNoteDto): Boolean {
        Log.d(TAG, "Removing my note:" + mynote.verseRange)
        return db.delete(MyNoteDatabaseDefinition.Table.MYNOTE, MyNoteColumn._ID + "=" + mynote.id, null) > 0
    }

    val allMyNotes: List<MyNoteDto>
        get() {
            Log.d(TAG, "about to getAllMyNotes")
            val allMyNotes: MutableList<MyNoteDto> = ArrayList()
            val c = db.query(builder(MyNoteQuery.TABLE).columns(MyNoteQuery.COLUMNS).create())
			c.use { c ->
				if (c.moveToFirst()) {
					while (!c.isAfterLast) {
						val mynote = getMyNoteDto(c)
						allMyNotes.add(mynote)
						c.moveToNext()
					}
				}
			}
            Log.d(TAG, "allMyNotes set to " + allMyNotes.size + " item long list")
            return allMyNotes
        }

    fun getMyNotesInBook(book: BibleBook): List<MyNoteDto> {
        Log.d(TAG, "about to getMyNotesInPassage:" + book.osis)
        val notesList: MutableList<MyNoteDto> = ArrayList()
        val c = db.query(builder(MyNoteQuery.TABLE).columns(MyNoteQuery.COLUMNS).selection(MyNoteColumn.KEY + " LIKE ?", arrayOf(book.osis + ".%")).create())
		c.use { c ->
			if (c.moveToFirst()) {
				while (!c.isAfterLast) {
					val mynote = getMyNoteDto(c)
					notesList.add(mynote)
					c.moveToNext()
				}
			}
		}
        Log.d(TAG, "myNotesInPassage set to " + notesList.size + " item long list")
        return notesList
    }

    private fun getMyNoteDto(id: Long): MyNoteDto? {
        var mynote: MyNoteDto? = null
        val c = db.query(builder(MyNoteQuery.TABLE).columns(MyNoteQuery.COLUMNS).selection(MyNoteColumn._ID + "=?", arrayOf(id.toString())).create())
		c.use { c ->
			if (c.moveToFirst()) {
				mynote = getMyNoteDto(c)
			}
		}
        return mynote
    }

    /**
     * Find bookmark starting at the location specified by key.
     * If it starts there then the initial part of the key should match.
     */
    fun getMyNoteByStartVerse(startVerse: String): MyNoteDto? {
        var mynote: MyNoteDto? = null
        var c: Cursor? = null
        try { // exact match
            c = db.query(builder(MyNoteQuery.TABLE).columns(MyNoteQuery.COLUMNS).selection(MyNoteColumn.KEY + "=?", arrayOf(startVerse)).create())
            if (!c.moveToFirst()) { // start of verse range
                c = db.query(builder(MyNoteQuery.TABLE).columns(MyNoteQuery.COLUMNS).selection(MyNoteColumn.KEY + " LIKE ?", arrayOf("$startVerse-%")).create())
                if (!c.moveToFirst()) {
                    return null
                }
            }
            mynote = getMyNoteDto(c)
        } finally {
            c!!.close()
        }
        return mynote
    }

    /** return Dto from current cursor position or null
     * @param c
     * @return
     * @throws NoSuchKeyException
     */
    private fun getMyNoteDto(c: Cursor?): MyNoteDto {
        val dto = MyNoteDto()
        try { //Id
            val id = c!!.getLong(MyNoteQuery.ID)
            dto.id = id
            //Verse
            val key = c.getString(MyNoteQuery.KEY)
            var v11n: Versification? = null
            if (!c.isNull(MyNoteQuery.VERSIFICATION)) {
                val v11nString = c.getString(MyNoteQuery.VERSIFICATION)
                if (!StringUtils.isEmpty(v11nString)) {
                    v11n = Versifications.instance().getVersification(v11nString)
                }
            }
            if (v11n == null) {
                Log.d(TAG, "Using default Versification")
                // use default v11n
                v11n = Versifications.instance().getVersification(Versifications.DEFAULT_V11N)
            }
            Log.d(TAG, "Versification found:$v11n")
            try {
                dto.verseRange = VerseRangeFactory.fromString(v11n, key)
            } catch (e: Exception) {
                Log.e(TAG, "Note saved with incorrect versification", e)
                // fix problem where KJV was always the v11n even for dc books
// NRSVA should contain most dc books and allow verse to be fetched
                val v11nWithDC = Versifications.instance().getVersification("NRSVA")
                dto.verseRange = VerseRangeFactory.fromString(v11nWithDC, key)
            }
            //Note
            val mynote = c.getString(MyNoteQuery.MYNOTE)
            dto.noteText = mynote
            //Update date
            val updated = c.getLong(MyNoteQuery.LAST_UPDATED_ON)
            dto.lastUpdatedOn = Date(updated)
            //Create date
            val created = c.getLong(MyNoteQuery.CREATED_ON)
            dto.createdOn = Date(created)
        } catch (nke: NoSuchKeyException) {
            Log.e(TAG, "Key error", nke)
        }
        return dto
    }

    private interface MyNoteQuery {
        companion object {
            const val TABLE = MyNoteDatabaseDefinition.Table.MYNOTE
            val COLUMNS = arrayOf(MyNoteColumn._ID, MyNoteColumn.KEY, BookmarkColumn.VERSIFICATION, MyNoteColumn.MYNOTE, MyNoteColumn.LAST_UPDATED_ON, MyNoteColumn.CREATED_ON)
            const val ID = 0
            const val KEY = 1
            const val VERSIFICATION = 2
            const val MYNOTE = 3
            const val LAST_UPDATED_ON = 4
            const val CREATED_ON = 5
        }
    }

    companion object {
        private const val TAG = "MyNoteDBAdapter"
    }
}
